feature: custom domains using Fairings

igalic opened the pull request:
This Pull Request addresses #94; replaces #447;
We start by adding `custom_domain` to the `blogs` table; and a custom type Host(String)

A blog with `custom_domain` has special routes with higher priority, but administrative routes will remain on the instance host.

We'll use `Fairings` to extract the `Host` header, to then specially route them.

Domain administration is left to the user (and/or admin)
(but we might be able to provide a check on whether it's already correctly configured)

-------

## TODO

- [x] PostgreSQL Migrations
- [x] SQLite Migrations
- [x] Custom time to represent domain (`Host`)
- [x] Expand `Blog` (and `NewBlog`)
- [x] Update `/blog/new` and `/blog/edit` to add custom domain field
- [x] Cache custom domains on startup and on insert/update.
- [x] `/custom_domains/` mount-point which handles these domains
- [ ] implement routes for
    - [x] blog details
    - [x] post details 
    - [x] blog activity_details
    - [ ] search
    - [ ] routes necessary for federation??
- [ ] implement redirect for the "legacy" routes for the above
    - [x] blog details
    - [x] post details
    - [ ] blog activity_details ??
    - [ ] search
- [x] Update URLs in templates
- [ ] Update URLs in base template for search
- [ ] Check correct domain setup
- [ ] Test: Federation
- [ ] Some explanations (in the UI, next to the setting and in the docs)

igalic committed:
add custom_domain (default to NULL) to schema

igalic committed:
add Host(String) wrapper type

we can use this to handle rocket Requests (from_request)
but we still have to figure out how to automatically assign the type to
custom_domain: Optional<Host>, and public_domain: Host.

igalic committed:
implement Host more completely by doing less

we now use DieselNewType and Shrinkwrap to automatically derive all the
things we need.

igalic committed:
list custom_domains

thanks a lot to @fdb-hiroshima and @BaptisteGelez for helping with the
code.

fdb-hiroshima commented on src/main.rs:
```diff
@@ -156,6 +156,9 @@ Then try to restart Plume
     }
 
+    let custom_domains = plume_models::blogs::Blog::list_custom_domains(&dbpool.get().unwrap()).unwrap();
+    dbg!(custom_domains);
```
Although it's very useful for debugging and fine for now, I'm gonna point this just to make sure we don't forget it when time for merging come

fdb-hiroshima commented on plume-models/src/blogs.rs:
```diff
@@ -311,6 +311,15 @@ impl Blog {
+            .select(blogs::custom_domain)
+            .load::<Option<String>>(conn)
+            .map_err(Error::from)
+            .map(|res| res.into_iter().map(|s| s.unwrap()).collect::<Vec<_>>())
```
```
error: redundant closure found
   --> plume-models/src/blogs.rs:321:44
    |
321 |             .map(|res| res.into_iter().map(|s| s.unwrap()).collect::<Vec<_>>())
    |                                            ^^^^^^^^^^^^^^ help: remove closure as shown: `std::option::Option::unwrap`
    |
    = note: `-D clippy::redundant-closure` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

igalic committed:
create & attach an AdHoc Fairing for dealing with Custom Domains

this rewrites the URL to /custom_domain/<url>
(we have no route handlers for this path yet)

Lots of help from @fdb-hiroshima & @BaptisteGelez in dealing with the
borrow checker — thank you 💜️

igalic committed:
appease clippy

igalic committed:
cache custom_domains list

follow #572 and cache the list of custom domains.

igalic committed:
normalize URLs before setting them

igalic committed:
move impl closer to mother struct

igalic committed:
CustomDomainFairing must not match /static/ and /api/

thanks, again, @fdb-hiroshima for helping with this code, when i got
stuck in lifetime-hecc.

igalic committed:
Add custom_details and custom_activity_details as first routes

how this works: we use find_by_host() to find the Host in question, the
defer to the existing function!

Caveat: Currently, we, in that function, we do another lookup DB lookup
for the Blog, even thou we already know it.
It might be wise, to have both of those another wrapper here?!

fdb-hiroshima commented on plume-models/src/blogs.rs:
```diff
@@ -195,6 +195,23 @@ impl Blog {
+            .load::<Blog>(&*c.conn)?
+            .into_iter()
+            .next();
+        if let Some(from_db) = from_db {
```
you might wanna check [ok_or](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.ok_or)

fdb-hiroshima commented on plume-models/src/blogs.rs:
```diff
@@ -195,6 +195,23 @@ impl Blog {
+        if let Some(from_db) = from_db {
+            Ok(from_db)
+        } else {
+            // we should never get here, because
```
(this is actually false, load can return a success with zero elements (type `select * from blogs where custom_domain = 'none';` in an sql shell if you need to convince yourself, no error, but nothing matched))

igalic replied:
i understand, but this function is only called after our `guard` already put us here

igalic replied:
removed comment.

fdb-hiroshima commented on plume-models/src/blogs.rs:
```diff
@@ -195,6 +195,23 @@ impl Blog {
+            .limit(1)
+            .load::<Blog>(&*c.conn)?
+            .into_iter()
+            .next();
```
I know this not used everywhere else, but using [first](https://docs.diesel.rs/diesel/query_dsl/trait.RunQueryDsl.html#method.first) would be more meaning full

igalic replied:
`limit(1)` also doesn't make sense given that `custom_domain` is `UNIQUE`

fdb-hiroshima replied:
agreed. If I'm correct however, first returns a single entry, not a Vec where you have to try to take the first

igalic replied:
```rust
    pub fn find_by_host(c: &PlumeRocket, host: Host) -> Result<Blog> {
        let from_db = blogs::table
            .filter(blogs::custom_domain.eq(host))
            .load::<Blog>(&*c.conn)?
            .first();
        from_db.ok_or(Error::NotFound)
    }
```
results in
```
error[E0308]: mismatched types
   --> plume-models/src/blogs.rs:203:9
    |
198 |     pub fn find_by_host(c: &PlumeRocket, host: Host) -> Result<Blog> {
    |                                                         ------------ expected `std::result::Result<blogs::Blog, Error>` because of return type
...
203 |         from_db.ok_or(Error::NotFound)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `blogs::Blog`, found reference
    |
    = note: expected type `std::result::Result<blogs::Blog, _>`
               found type `std::result::Result<&blogs::Blog, _>`

error: aborting due to previous error
```
i don't know if to laugh or cry.

fdb-hiroshima replied:
I meant [diesel first](https://docs.diesel.rs/diesel/query_dsl/trait.RunQueryDsl.html#method.first), in place of load. Not [Vec first](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.first) (actually it's slice's, but Vec deref to slice)

igalic replied:
```
error[E0308]: mismatched types
   --> plume-models/src/blogs.rs:199:9
    |
198 |       pub fn find_by_host(c: &PlumeRocket, host: Host) -> Result<Blog> {
    |                                                           ------------ expected `std::result::Result<blogs::Blog, Error>` because of return type
199 | /         blogs::table
200 | |             .filter(blogs::custom_domain.eq(host))
201 | |             .first::<Blog>(&*c.conn)
    | |____________________________________^ expected enum `Error`, found enum `diesel::result::Error`
    |
    = note: expected type `std::result::Result<_, Error>`
               found type `std::result::Result<_, diesel::result::Error>`
```

but if i use `QueryResult<Blog>`, then i can't use `?` for error handling later 🙍

igalic replied:
actually, `rustc` is saying exactly what i need to do, so i did that.

igalic committed:
replace if / else with ok_or()

igalic committed:
simplify retrieval in find_by_host()

we can use "first()" instead of limit(1).load().etc…
since on a UNIQUE field, we only expect 1 result.

first() returns QueryResult, which is Result<T, diesel::Error>, so we
need to implement a converter for that error type.

This commit addresses @fdb-hiroshima's review.

igalic committed:
extract common routing code into private "_guts()" functions

BaptisteGelez commented on plume-models/src/blogs.rs:
```diff
@@ -195,14 +195,10 @@ impl Blog {
-            .into_iter()
-            .next();
-        from_db.ok_or(Error::NotFound)
+            .first::<Blog>(&*c.conn)
```
What about adding a `map_err(|_| Error::NotFound)` here, to keep `Result<Blog>` as the return type (with the advantages it has, like having the possibility to "throw" it to get an error page)?

BaptisteGelez replied:
And also it would allow you to remove the `From<diesel::result::Error> for ErrorPage` impl.

fdb-hiroshima replied:
> What about adding a `map_err(|_| Error::NotFound)` here

or `map_err(Error::from)` ?

igalic replied:
seemed like a good deal to me to directly map a `diesel::result::Error` to an `ErrorPage`
otoh, i can also see how this would lead to the kind of error handling that PHP + MySQL are famous for

igalic committed:
simplify / unify error handling

We want to return a Result<Blog>, instead of a QueryResult<Blog>.
We can use map_err() to map it to the desired error type.

thanks to review from @BaptisteGelez & @fdb-hiroshima.

igalic committed:
extend 404 handler to handle all the requests our custom_ routes dont

igalic commented:
so: I'm a bit at an impasse or crossroads with this patch, and it all kinda depends on how far i wanna push it.

For now, i've added a 404 handler, which redirects to the main-site, which means, actually, everything works now 😹 (by redirecting to the main site)

We could present the complete illusion that this custom domain, is the real domain, and all there is is this one blog.

To do this, I would have to adapt the URLs `ap_url`, `inbox_url` and `outbox_url` on blog creation (or update) and i would have to modify the templates to use the correct URL scheme.
Alternatively we create an URL builder which we resolves the correct URLs.

Currently, there are a lot of places in which URLs are built ad-hoc.
In some cases, this is okay, because it's just pointing to a rocket route, but in many cases it feels painful, because we're doing manual labour that a function could remove.

the `compute_box()` function for instance, feels quite pointless. There is no computation happening at all, just concatenation! `compute_box()` itself calls `ap_url()` which only prepends `https://`

I would like to propose a Universal URL builder… or rather, a plumey-versal URL builder, which can be used from any part of plume's code to construct a correct plume URL.

fdb-hiroshima commented:
> To do this, I would have to adapt the URLs ap_url, inbox_url and outbox_url on blog creation (or update) and i would have to modify the templates to use the correct URL scheme.

I think it would be best to still handle most of (~all) ap stuff on the main domain, having an id (as requested in #296) for the activity that goes to the main domain, with a redirect to the correct path when queried (so either the main domain with better name than the id, or custom one when applicable). Also with some webfinger trick it should be possible to have more than one handle for a blog, so there could be the one from the main domain, that shouldn't change because it would break a lot of federation stuff, and an alias based on the custom domain. All this would be in the spirit of not breaking things if the custom domain for a blog changes

> I would like to propose a Universal URL builder… or rather, a plumey-versal URL builder, which can be used from any part of plume's code to construct a correct plume URL.

that would be a good idea for everything that need to integrate the domain in addition to a path. It could be a small macro that call the rocket one, and prepand the domain name, either from local_instance if none is given, or the one given (which would be a custom domain)

igalic committed:
Allow searching from custom_domain

igalic commented:
okay, so, at this point, i'm about as done as i can be, without touching the templates or refactoring the URL building.

igalic committed:
add url to Blog, this seems useful.

rfwatson commented on migrations/sqlite/2019-05-22-124153_add_custom_domains/up.sql:
```diff
@@ -0,0 +1,57 @@
+-- add custom_domain to blogs
+CREATE TABLE IF NOT EXISTS "blogs_add_custom_domain" (
```
Just wondering why copying the table is required, instead of just adding the column?

igalic replied:
because [SQLite supports a limited subset of ALTER TABLE](https://www.sqlite.org/lang_altertable.html)

rfwatson replied:
Damn, I missed it was SQLite. Thanks.

igalic committed:
redirect blog urls to custom_domain if it exists

BaptisteGelez commented on src/routes/blogs.rs:
```diff
@@ -35,7 +37,63 @@ pub fn details(name: String, page: Option<Page>, rockets: PlumeRocket) -> Result
+
+    match (blog.custom_domain, page) {
+        (Some(ref custom_domain), Some(ref page)) => {
+            Ok(Redirect::to(format!("https://{}/?{}", custom_domain, page)).into())
```
I think it will not properly format the page: you will get something like `https://doma.in/?2` when you probably want `?page=2`.

elegaanz reviewed (commented):
I gave it a try, and it seems to work quite well so far! I left a few comment about the code itself.

igalic committed:
fix issues pointed out by @BaptisteGelez in review

igalic committed:
custom_domainify posts::details

BaptisteGelez commented on src/routes/posts.rs:
```diff
@@ -31,28 +31,14 @@ use routes::{
-    slug: String,
+fn detail_guts(
+    blog: &Blog,
+    post: &Post,
```
Does that (and the `blog: &Blog` above) work?

BaptisteGelez replied:
nvm, I thought it was a route, but it is just a regular function.

BaptisteGelez commented on src/routes/posts.rs:
```diff
@@ -94,7 +80,79 @@ pub fn details(
+
+    match (blog.custom_domain, responding_to) {
+        (Some(ref custom_domain), Some(ref responding_to)) => Ok(Redirect::to(format!(
+            "https://{}/?responding_to={}",
```
I think this URL (and the one after) should include the article's slug.

igalic replied:
oh gosh yes that was a copy paste!

fdb-hiroshima commented on src/routes/posts.rs:
```diff
@@ -94,7 +80,79 @@ pub fn details(
+        (Some(ref custom_domain), _) => {
+            Ok(Redirect::to(format!("https://{}/", custom_domain)).into())
+        }
+        (None, _) => panic!("This code path should have already been handled!"),
```
You can use the handy `unreachable!("<msg>")` instead of a plain panic, it has the same effect but is probably clearer

igalic committed:
follow up on @fdb-hiroshima & @BaptisteGelez review

trinity-1686a committed:
add url! macro for custom domain path uri creation

see doc-comment for limitations

igalic committed:
move custom_ route functions into a custom namespace

this way, we can actually use the url! macro

trinity-1686a committed:
allow more syntaxes for url

allow for params that are used only by one of normal/custom routes

igalic committed:
fix syntax: we're now down to type errors

trinity-1686a committed:
make url! work better

there is still some issue with `None`, and an error making no sense at
some place

trinity-1686a committed:
finally fix url!

also please a bit clippy

igalic committed:
start modifying /blogs/new to add custom_domain to the form

elegaanz committed:
Strings have to be cloned

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -90,6 +155,12 @@ pub fn create(form: LenientForm<NewBlogForm>, rockets: PlumeRocket) -> RespondOr
     let intl = &rockets.intl.catalog;
     let user = rockets.user.clone().unwrap();
 
+    let custom_domain = if *(&form.custom_domain.is_empty()) {
```
clippy says 
```suggestion
    let custom_domain = if form.custom_domain.is_empty() {
```

AnaGelez replied:
I can't accept your proposal via Github (apparently I don't have the right to push to this repository :upside_down_face:), but I made it manually, thanks.

igalic committed:
fix clippy warning

Co-Authored-By: fdb-hiroshima <35889323+fdb-hiroshima@users.noreply.github.com>

igalic commented on src/routes/blogs.rs:
```diff
@@ -72,6 +136,7 @@ pub fn new_auth(i18n: I18n) -> Flash<Redirect> {
 pub struct NewBlogForm {
     #[validate(custom(function = "valid_slug", message = "Invalid name"))]
     pub title: String,
```
we can add a validate function here, but it seems rather limited what we can return

for instance: DNS doesn't resolve? we can't say anything at this point, but should still tell the person that it doesn't resolve (yet)

another test would then be to try to connect and see if it's connecting to us, but the question here is: how??

igalic replied:
would https://github.com/Plume-org/Plume/issues/296 be a solution to this issue?

AnaGelez replied:
I think we should just make a DNS request to check that there is a CNAME record pointing to the main domain? And if it doesn't resolve, we can probably show a warning like: "Your custom domain is not properly configured yet, please re-try later in your blog settings." (or something like that, I trust you to find a better wording).

And I don't think #296 is a solution to this specific issue, even if it may help somewhere else in this branch…

fdb-hiroshima replied:
In some complex setup, checking only a CNAME can be not enough, for instance at home I have `subdomain.my_domain` => `my_domain`, but `my_domain` is also a CNAME to a domain provided by my isp. Suppose Plume run on subdomain.my_domain, pointing to `my_domain`, or A/AAAA to my public ips would work, but would also be rejected. I think having an endpoint which we can query with Reqwest is more solid to complex/unexpected setup

AnaGelez replied:
And it avoids having to do some DNS queries directly, so even better :smile:

igalic replied:
what would we be looking for when making such a Reqwest?
how do we identify our own instance?

fdb-hiroshima replied:
We could query `/nodeinfo/2.0` and metadata.nodeName matches local value. It does not protect against voluntary "mistake" (actually, nothing can), but will catch configuration error

igalic replied:
is there a cryptographically secure  way in ActivityPub or OcapPub to do this?

fdb-hiroshima replied:
I don't know OcapPub, AP has no cryptography (non normative security sections everywhere), but in general, no cryptographic protection has any effect against a proxy attack, someone could just temporary proxy traffic to Plume, then do nothing else (that's a big problem on smart-cards btw, they tend to use timing info to fight against, but it's not very efficient). The errors we can catch here are mistakes, not clever enough attacks

igalic reviewed (commented):
validation?

elegaanz committed:
Basic domain validation

The actual DNS check still needs to be written

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -136,11 +209,19 @@ pub fn create(form: LenientForm<NewBlogForm>, rockets: PlumeRocket) -> RespondOr
+        )
+        .into()
+    } else {
+        Flash::warn(
```
```suggestion
        Flash::warning(
```

igalic committed:
review by @fdb-hiroshima

Co-Authored-By: fdb-hiroshima <35889323+fdb-hiroshima@users.noreply.github.com>

igalic committed:
start exploring some ideas for domain validation

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -95,6 +95,11 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
 }
 
+#[get("/validate/<validation_id>")]
+pub fn validate(validation_id: String, rockets: PlumeRocket) -> Result {
```
the name mismatch the one in main.rs, and Result take 2 type arguments (probably both String I think)

igalic replied:
thanks / sorry for the noise 🙇‍♀️

the two files were edited with about a week or two in between 😅

fdb-hiroshima replied:
np :wink:

igalic committed:
implement domain validation using rocket::State

this doesn't work (yet?), because i don't know how to store something
mutable in State<>

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -96,8 +98,27 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
+    let now = Instant::now();
+
+    // nope, expired (410: gone)
+    if now.checked_duration_since(*valid_until).unwrap().as_secs() > 0 {
```
you should use `duration_since` instead, as you're unwrapping it it's doing exactly the same thing, and it's stable

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -96,8 +98,27 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
+    // nope, expired (410: gone)
+    if now.checked_duration_since(*valid_until).unwrap().as_secs() > 0 {
+        valid_domains.inner().remove(validation_id.as_str());
+        return Err(String::from("410"));
```
(this will only return the string 410, not the error code)

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -96,8 +98,27 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
+) -> Result<String, String> {
+    let value = valid_domains.inner().get(validation_id.as_str());
+    if value.is_none() {
+        return Err(String::from("404"));
```
(this will only return the string 404, not the error code)

fdb-hiroshima commented on src/main.rs:
```diff
@@ -196,6 +197,7 @@ Then try to restart Plume
         }
     });
 
+    let mut valid_domains: HashMap<&str, Instant> = HashMap::new();
```
I'd be in favor of using the newtype design pattern for this, so that it make more sense what it is, even at the type level.
I don't know where are coming the &str from, but you might want to use String instead, or you'll have some difficulties with lifetime

igalic replied:
aye. i haven't gotten that far yet, because i haven't implemented the other side of this

trinity-1686a reviewed (commented):
:eyes:

igalic committed:
correctly mutex-lock and modify valid_domains

thanks to @fdb-hiroshima for the review

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -100,25 +101,29 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
+    let value = validation_getter.get(validation_id.as_str());
     if value.is_none() {
-        return Err(String::from("404"));
+        return Status::new(404, "validation id not found");
```
I believe there are consts that already contains standards http status in the module

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -100,25 +101,29 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
-    if now.checked_duration_since(*valid_until).unwrap().as_secs() > 0 {
-        valid_domains.inner().remove(validation_id.as_str());
-        return Err(String::from("410"));
+    if now.duration_since(*valid_until).as_secs() > 0 {
```
You should remove the feature, as it's no longer used

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -100,25 +101,29 @@ pub fn new(rockets: PlumeRocket, _user: User) -> Ructe {
-        return Err(String::from("410"));
+    if now.duration_since(*valid_until).as_secs() > 0 {
+        validation_map.remove(validation_id.as_str());
+        return Status::new(410, "validation expired");
```
Same as above

igalic replied:
```
error[E0308]: mismatched types
   --> src/routes/blogs.rs:113:28
    |
113 |         return Status::new(Status::NotFound, "validation id not found");
    |                            ^^^^^^^^^^^^^^^^ expected u16, found struct `rocket::http::Status`
    |
    = note: expected type `u16`
               found type `rocket::http::Status`

error[E0308]: mismatched types
   --> src/routes/blogs.rs:123:28
    |
123 |         return Status::new(Status::Gone, "validation expired");
    |                            ^^^^^^^^^^^^ expected u16, found struct `rocket::http::Status`
    |
    = note: expected type `u16`
               found type `rocket::http::Status`
```

fdb-hiroshima replied:
Have you tried simply
```suggestion
        return Status::Gone;
```
?

igalic replied:
No, cuz I wanted to give a specific response.

fdb-hiroshima replied:
```suggestion
        return Custom(Status::Gone, "validation expired");
```
imported from rocket::response::status::Custom
Your version is sending a standard error code with a non standard text, some implementations don't handle this well. This return the corresponding "error" text, and set the body of the result to the text you provide

trinity-1686a reviewed (commented):
:eyes:

igalic committed:
one step closer to validating custom domains!

igalic committed:
validate custom domain!

fdb-hiroshima commented on src/routes/blogs.rs:
```diff
@@ -179,8 +179,24 @@ fn valid_slug(title: &str) -> Result<(), ValidationError> {
+        Instant::now().checked_add(Duration::new(60, 0)).unwrap(),
+    );
+
+    true
```
I don't get returning a constant value, I don't see any early return?

igalic committed:
do not panic if reqwest does not work

igalic committed:
fix getter router for domain_validation!

igalic committed:
also mount domain_validation into /custom_domains/

igalic committed:
Allow editing blogs with custom_domain

n.b.: no validation yet

igalic committed:
use Custom status code

thanks, @fdb-hiroshima

Co-Authored-By: fdb-hiroshima <35889323+fdb-hiroshima@users.noreply.github.com>

igalic committed:
use Status::* as @fdb-hiroshima suggested

igalic committed:
when building links, skip both, /custom_domain/ and <custom_domain>

igalic committed:
fix mount point for blog post routing function

igalic committed:
restore df7adaf0e067802daf3e83935cf9140fb15ca17b ← after fixing mount-point; fix query() parsing

igalic committed:
Allow updating of custom_domain

AnaGelez commented on src/routes/blogs.rs:
```diff
@@ -351,6 +351,10 @@ pub fn edit(name: String, rockets: PlumeRocket) -> Result<Ructe, ErrorPage> {
         let medias = Media::for_user(conn, user.id).expect("Couldn't list media");
+        let custom_domain = match blog.custom_domain {
+            Some(ref c) => c.to_string(),
+            _ => String::from(""),
```
It is a quite personal opinion, but I think `blog.custom_domain.map(ToString::to_string).unwrap_or_default()` is more explicit (but, yes, it is only my opinion, maybe it is not).

igalic replied:
another question might be: why were you able to use `Option<String>` for `theme` in `EditForm` but i have to use `String` here?
